db/blip_sync_context.go

```diff
@@ -314,40 +320,44 @@ func (bsc *BlipSyncContext) sendRevisionWithProperties(sender *blip.Sender, docI
 	}
 	bsc.dbStats.StatsDatabase().Add(base.StatKeyNumDocReadsBlip, 1)
 
-	base.Tracef(base.KeySync, "Sending revision %s/%s, body:%s, properties: %v, attDigests: %v", base.UD(docID), revID, base.UD(string(bodyBytes)), base.UD(properties), attDigests)
+	base.TracefCtx(bsc.loggingCtx, base.KeySync, "Sending revision %s/%s, body:%s, properties: %v, attDigests: %v", base.UD(docID), revID, base.UD(string(bodyBytes)), base.UD(properties), attDigests)
 
 	// TODO: When CBG-881 is implemented, update this stat and SendRevCountError based on sendRev response
 	bsc.replicationStats.SendRevCount.Add(1)
 
-	if len(attDigests) > 0 {
+	// asyncronously wait for a response if we have attachment digests to verify, or if we have a registered callback.
+	verifyAttachments := len(attDigests) > 0
+	awaitResponse := responseCallback != nil || verifyAttachments
+
+	if verifyAttachments {
 		// Allow client to download attachments in 'atts', but only while pulling this rev
 		bsc.addAllowedAttachments(attDigests)
-		if !bsc.sendBLIPMessage(sender, outrq.Message) {
-			return ErrClosedBLIPSender
-		}
+	}
+
+	if !awaitResponse {
+		outrq.SetNoReply(true)
+	}
+
+	if !bsc.sendBLIPMessage(sender, outrq.Message) {
+		return ErrClosedBLIPSender
+	}
+
+	if awaitResponse {
 		go func() {
 			defer func() {
 				if panicked := recover(); panicked != nil {
 					base.Warnf("[%s] PANIC handling 'sendRevision' response: %v\n%s", bsc.blipContext.ID, panicked, debug.Stack())
 					bsc.Close()
 				}
 			}()
-			defer bsc.removeAllowedAttachments(attDigests)
-			outrq.Response() // blocks till reply is received
-			base.Tracef(base.KeySync, "Received response for sendRevisionWithProperties rev message %s/%s", base.UD(docID), revID)
-		}()
-	} else {
-		outrq.SetNoReply(true)
-		if !bsc.sendBLIPMessage(sender, outrq.Message) {
-			return ErrClosedBLIPSender
-		}
-	}
 
-	if response := outrq.Response(); response != nil {
-		if response.Type() == blip.ErrorType {
-			errorBody, _ := response.Body()
-			base.WarnfCtx(bsc.loggingCtx, "Client returned error in rev response for doc %q / %q: %s", base.UD(docID), revID, errorBody)
-		}
+			resp := outrq.Response() // blocks till reply is received
+			base.TracefCtx(bsc.loggingCtx, base.KeySync, "Received response for sendRevisionWithProperties rev message %s/%s", base.UD(docID), revID)
+			if verifyAttachments {
+				bsc.removeAllowedAttachments(attDigests)
+			}
+			responseCallback(resp)
```

Comment:
Do we need to pass around a responseCallback?  Or can we just do "if bsc.postSendRevisionResponseCallback..." here and execute the code you've got in the callback?

Reply:
Yeah that simplifies, will just need to pass `seq` down into one code-path that doesn't have it yet.